external-builds/lib_python/repo_management.py

```diff
@@ -0,0 +1,407 @@
+#! python
+
+import argparse
+import shlex
+import shutil
+import subprocess
+import sys
+import os
+import glob
+import platform
+from pathlib import Path, PurePosixPath
+from urllib.parse import urlparse, urlunparse, quote
+
+import subprocess
+
+TAG_UPSTREAM_DIFFBASE = "THEROCK_UPSTREAM_DIFFBASE"
+TAG_HIPIFY_DIFFBASE = "THEROCK_HIPIFY_DIFFBASE"
+HIPIFY_COMMIT_MESSAGE = "DO NOT SUBMIT: HIPIFY"
+
+class RockProjectRepo():
+    def __init__(self,
+                project_name,
+                project_src_dir,
+                project_build_dir,
+                project_repo_url,
+                project_version_hashtag):
+        self.project_name = project_name
+        self.project_src_dir = Path(project_src_dir)
+        self.project_build_dir = Path(project_build_dir)
+        self.project_repo_url = project_repo_url
+        self.project_version_hashtag = project_version_hashtag
+        ROCK_BUILDER_HOME_DIR = Path(os.environ["ROCK_BUILDER_HOME_DIR"])
+        self.project_patch_dir_base = ROCK_BUILDER_HOME_DIR / "patches" / project_name
+        os.environ["ROCK_BUILDER_APP_SRC_DIR"] = str(project_src_dir)
+
+    # private methods
+    def __exec_subprocess_cmd(self, exec_cmd):
+        ret = True
+        if exec_cmd is not None:
+            print("exec_cmd")
+            # capture_output=True --> can print output after process exist, not possible to see the output during the build time
+            # capture_output=False --> can print output only during build time
+            #result = subprocess.run(exec_cmd, shell=True, capture_output=True, text=True)
+            result = subprocess.run(exec_cmd, shell=True, capture_output=False, text=True)
+            if result.returncode == 0:
+                print(result.stdout)
+                ret = False
+            else:
+                print(result.stdout)
+                print(f"Error: {result.stderr}")
+        return ret
+
+    def __exec_subprocess_batch_file(self, batch_file):
+        ret = True
+        if batch_file is not None:
+            print("batch_file: " + batch_file)
+            # capture_output=True --> can print output after process exist, not possible to see the output during the build time
+            # capture_output=False --> can print output only during build time
+            #result = subprocess.run(exec_cmd, shell=True, capture_output=True, text=True)
+            result = subprocess.run([batch_file], shell=True, capture_output=False, text=True)
+            if result.returncode == 0:
+                print(result.stdout)
+                ret = False
+            else:
+                print(result.stdout)
+                print(f"Error: {result.stderr}")
+        return ret
+
+    def __replace_env_variables(self, cmd_str):
+        ret = os.path.expandvars(cmd_str)
+        #print(ret)
+        return ret
+
+    def __get_latest_file(self, path, extension=None):
```

Comment:
Prefer to mark types for arguments and add docstrings when the function name is not enough to understand what a function does. Is `path` a `Path` or `str`?